RzUtil: softfloat support based on bitvector

Heersin opened the pull request:
# PLEASE SQUASH ME

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**
- still work in progress basic IEEE754 float support in rizin util lib. (#2005 RZIL float support will be based on this according to thestr4ng3r's advice)
- for earlier code review
- take softfloat3 source code as a reference (http://www.jhauser.us/arithmetic/SoftFloat.html)

**Test plan**
1. basic testcases (current)
2. Add random testcases generated by softfloat3

**Current problem (to-do plan)**
1. the physical implementations of C type are different on multiple real-world machines. We should find a solution to deal with such differences. (`new_from_single` etc.)
2. update according to latest BAP

**Closing issues**

Heersin committed:
Init float file in rz_util

Heersin committed:
Add test for float/double in C to bv

Heersin committed:
Update internal float

Heersin committed:
Implement some basic funcs, prepare to refactor float

Heersin committed:
Refactor float basic struct

Pass simple test

Hit problem in some magic operations, ready to take stupid emulate way

As-is float_add part done

Add propagate NaN

Add round

Discard old implementation, ready for new one

Heersin committed:
Move float project path

Heersin committed:
Add float add and sub

Add float mul, div and rem

Add round

Add detect_halfway and some documents

Fix bug and add fused_mul_add

Add special handler for inf/zero/nan

Heersin committed:
Compile float_test for testing util/float

Heersin committed:
Test float, waiting for merge bv fix

Heersin committed:
Fix bug in float_add and add some tests

Elimate some warnings

Add another test for float add, todo : float test functions and more test cases

Fix bug in sub float related and add test for fsub

Add 2 cases for add and sub

Add subnormal test case and fix bug in sub

Add mul testcases

Add div testcases

Clang-format code

Clean impl of float_rem

Add sqrt and muladd fabs basic

Add match macro

Remove match macro design and add inf/nan process for basic operations

Add trunc and rem

Fix code

Clang-format

Add trunc test

Add abs test

Add new_from_hex and test

Fix bug in fused mul add and add test

Add one more test for fma and bug fixing

Add one more test for fma and bug fixing

Add simple round test

Add special num test for add/sub

Fix bug in special num and add related tests

Bug fixing and add rem test, todo solve the precision in rem

Solve the precision lost in rem

Add a failed test cases to solve

Add sqrt test cases

XVilka commented on librz/include/rz_util/rz_float.h:
```diff
@@ -0,0 +1,88 @@
+#ifndef RZ_FLOAT_H
```
Missing SPDX

XVilka commented on librz/include/rz_util/rz_float.h:
```diff
@@ -0,0 +1,88 @@
+ * IEEE 754 only for now
+ */
+typedef enum float_format_enum {
+	RZ_FLOAT_IEEE754_BIN_32,
```
Could you please add Doxygen comments to these members too?

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+#include "float_internal.c"
```
SPDX

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+#include "float_internal.c"
+
+/**
+ * This file implements IEEE-754 binary float number (32/64/128)
```
Maybe use `\file` here?

XVilka replied:
I also recommend to add a short description what are those different formats are - such as "single" float actively used below.

Heersin replied:
yep it's a good idea to add some description here. and also in pr, I will add them after completing the code

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+
+/**
+ * Get const attribute from float
+ * @param format RzFloatFormat, format of a float
```
Please use `\param` syntax

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+	case RZ_FLOAT_IEEE754_DEC_64:
+	case RZ_FLOAT_IEEE754_DEC_128:
+	default:
+		RZ_LOG_ERROR("TODO");
```
Probably `rz_warn_if_reached()` is more appropriate here

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+
+/**
+ * Destroy the float structure
+ * @param f float
```
Same and below, `\param`

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+ * @param value value of type `float`
+ * @return true if success
+ */
+RZ_API bool rz_float_set_from_single(RZ_NONNULL RzFloat *f, float value) {
```
This one should be carefully thought out since in C `float` could be different a bit between implementations.

Heersin replied:
Yes, the implementations of C float could be different in different machines, and we should find an appropriate way to solve it before release.

wargio replied:
can you explain exactly what is the issue? the float and double size are standardized and when you use half/extended/quad format you need to use compiler extensions, but even in that case those are standardized.
i think for those smaller/bigger formats we can use just bitvectors to set it.

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+		return false;
+	}
+
+	rz_bv_set_from_bytes_le(
```
Probably could be still in one line

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+ * @param value value of type `double`
+ * @return true if success
+ */
+RZ_API bool rz_float_set_from_double(RZ_NONNULL RzFloat *f, double value) {
```
Same applies

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+	return f;
+}
+
+RZ_API RZ_OWN RzFloat *rz_float_new_from_single(float value) {
```
Missing Doxygen

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+	return f;
+}
+
+RZ_API RZ_OWN RzBitVector *rz_float_get_exponent_squashed(RZ_NONNULL RzFloat *f) {
```
Doxygen for these

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+	return get_sign(f->s, f->r);
+}
+
+RZ_API RzFloatSpec rz_float_detect_spec(RZ_NONNULL RzFloat *f) {
```
Missing Doxygen, also this function should be better tested, I think

Heersin replied:
TODO : testcases for `detec_spec`

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+	rz_return_val_if_fail(f, RZ_FLOAT_SPEC_NOT);
+
+	RzFloatSpec ret = RZ_FLOAT_SPEC_NOT;
+	RzBitVector *exp_squashed = get_exp_squashed(f->s, f->r);
```
Could these be NULL? It's better to check them

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+	return ret;
+}
+
+RZ_API bool rz_float_is_inf(RZ_NONNULL RzFloat *f) {
```
Doxygen

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+	free(bvstr);
+}
+
+static void print_float(RzFloat *f) {
```
Instead of these functions I recommend to make something like `rz_float_as_string()` and use it instead, will be useful even after debugging is finished.

Heersin replied:
ok, like `bv_as_stirng()`, current `print` only used in debugging quickly.

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+	return ret;
+}
+
+static RZ_OWN RzFloat *propagate_float_nan(RZ_NONNULL RzFloat *left, RzFloatSpec ltype, RZ_NONNULL RzFloat *right, RzFloatSpec rtype) {
```
While this internal function, I think it will greatly benefit from brief description

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+	ut32 total_len = rz_float_get_format_info(format, RZ_FLOAT_INFO_TOTAL_LEN);
+
+	/// Extract fields from num
+	RzBitVector *l_exp_squashed = get_exp_squashed(left->s, left->r);
```
Can all these be NULL?

Heersin replied:
it's possible

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+	return ret;
+}
+
+static RZ_OWN RzFloat *fadd_mag(RZ_NONNULL RzFloat *left, RZ_NONNULL RzFloat *right, bool sign, RzFloatRMode mode) {
```
Same, a description what this function does would be great

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+	return result;
+}
+
+/// for 32/64/128
```
Doxygen, stating it implements the operations, I also recommend to group them together using `\@{`

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+	RzFloat *result = NULL;
+
+	PROC_SPECIAL_FLOAT_START(left, right)
+	/**
```
Probably better to move this in the function description instead

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+
+RZ_API RZ_OWN RzFloat *rz_float_rem_ieee_bin(RZ_NONNULL RzFloat *left, RZ_NONNULL RzFloat *right, RzFloatRMode mode) {
+	PROC_SPECIAL_FLOAT_START(left, right)
+	/**
```
Same

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+
+/**
+ * calculate a + b * c, and then round the result
+ * @param a
```
You can write `calculate \p a + \p b * \p c, and round the result after` and remove these `@param`

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+		b_exp_val = 1 - shift_dist;
+		rz_bv_lshift(b_mantissa, shift_dist);
+	}
+
```
This function is too long, I recommend splitting it into multiple

Heersin replied:
I think comments take many lines and make it "huge".  I'm not sure if it would be better to split it.

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+
+RZ_API RZ_OWN RzFloat *rz_float_sqrt_ieee_bin(RZ_NONNULL RzFloat *n, RzFloatRMode mode) {
+	/// Use Newton method now, May Optimize
+	/// TODO: use epsilon or prev == cur to stop ?
```
Epsilon, probably

XVilka commented on librz/util/float/float_internal.c:
```diff
@@ -0,0 +1,440 @@
+	}
+}
+
+// TODO : move this to bitvector.c in the future
```
Yes

Heersin replied:
TODO : move this to bitvector lib after merged

XVilka commented on librz/util/float/float_internal.c:
```diff
@@ -0,0 +1,440 @@
+
+/**
+ * Get a bitvector representation of exponent, have the same length of parameter `bv`
+ * @param bv RzBitVector, the bitvector interpreted as float
```
`\param`

XVilka commented on librz/util/float/float_internal.c:
```diff
@@ -0,0 +1,440 @@
+
+	ut32 exp_len = rz_float_get_format_info(format, RZ_FLOAT_INFO_EXP_LEN);
+	ut32 man_len = rz_float_get_format_info(format, RZ_FLOAT_INFO_MAN_LEN);
+	RzBitVector *res = rz_bv_new(exp_len + man_len + 1);
```
Check if NULL

XVilka commented on librz/util/float/float_internal.c:
```diff
@@ -0,0 +1,440 @@
+
+	ut32 exp_len = rz_float_get_format_info(format, RZ_FLOAT_INFO_EXP_LEN);
+	ut32 man_len = rz_float_get_format_info(format, RZ_FLOAT_INFO_MAN_LEN);
+	RzBitVector *res = rz_bv_new(exp_len + man_len + 1);
```
Check if NULL

XVilka commented on librz/util/float/float_internal.c:
```diff
@@ -0,0 +1,440 @@
+	return ret;
+}
+
+static bool rz_make_fabs(RzFloat *f) {
```
Brief description please

XVilka commented on librz/util/float/float_internal.c:
```diff
@@ -0,0 +1,440 @@
+	rz_bv_set(f->s, f->s->len - 1, false);
+}
+
+static RzFloat *rz_half_float(RzFloat *f) {
```
Also description

XVilka reviewed (changes requested):
Looks like a solid work. Kudos!

DMaroo commented on librz/util/float/float_internal.c:
```diff
@@ -0,0 +1,440 @@
+}
+
+static bool rz_make_fabs(RzFloat *f) {
+	rz_bv_set(f->s, f->s->len - 1, false);
```
Should be
```c
return rz_bv_set(f->s, f->s->len - 1, false);
```

DMaroo commented on librz/util/float/float_internal.c:
```diff
@@ -0,0 +1,440 @@
+ * @param r_bits_bound ut32 boundary of round bits
+ * @return bool return true if significant should be rounded, else return false
+ */
+inline bool detect_should_round(RzBitVector *sig, ut32 r_bits_bound) {
```
Should be `static`, just `inline` function leads to a compile error.

Rot127 commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+	rz_return_val_if_fail(f, false);
+	rz_float_fini(f);
+
+	RzFloatFormat format = f->r;
```
In `rz_float_fini` the struct is memset to 0. So `RzFloatFormart` will always be `RZ_FLOAT_IEEE754_BIN_32` if I'm not mistaken.

Heersin replied:
Yep, I will add a TODO for it.

Rot127 commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+
+static RZ_OWN RzFloat *gen_zero_float(RzFloatFormat format) {
+	RzFloat *ret = rz_float_new(format);
+	rz_float_init(ret);
```
This `rz_float_init` is a duplicate. It was already called in `rz_float_new`.

Rot127 commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+
+	cp->r = f->r;
+	cp->s = rz_bv_dup(f->s);
+
```
Not quite sure about this: But I would also make a copy of `f->exception` although it might be very unlikely that `rz_loat_dup` is called after an exception was set.

Rot127 commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1288 @@
+	ut32 exp_end = exp_start + rz_float_get_format_info(format, RZ_FLOAT_INFO_EXP_LEN);
+	for (ut32 i = exp_start; i < exp_end; ++i) {
+		rz_bv_set(bv, i, true);
+	}
```
```c
ut32 exp_start = rz_float_get_format_info(format, RZ_FLOAT_INFO_MAN_LEN);
ut32 exp_end = exp_start + rz_float_get_format_info(format, RZ_FLOAT_INFO_EXP_LEN);
for (ut32 i = exp_start; i < exp_end; ++i) {
      rz_bv_set(bv, i, true);
}
```
Since this "set all exp bits" code appears at least three times it could be put in a inline function IMHO.

Heersin replied:
Agree. Thanks

Heersin replied:
I checked the implementations again. they all required the extracted attribute `exp_start` to set flag bit. and the this code snippet is seldom used in other parts. so I think we don't need an inline here. 
maybe an inline function such as `rz_bv_set_range(bitvector, start, end, val)` would be better IMHO

Rot127 replied:
Yeah sure. Wasn't so sure about the inline anyways. But just replacing the for loop with some kind of `rz_bv_set_range` already looks better to me.

Rot127 commented on librz/util/float/float_internal.c:
```diff
@@ -0,0 +1,440 @@
+#include "rz_util.h"
```
```suggestion
#include <rz_util.h>
```

Rot127 commented on librz/util/float/float_internal.c:
```diff
@@ -0,0 +1,440 @@
+	case RZ_FLOAT_INFO_BIAS:
+		return 127;
+	default:
+		return 0;
```
`rz_warn_if_reached()`? Or is there a legitimate reason I miss? Same below

Heersin replied:
ah, my mistake. `rz_warn_if_reached` should be added

Rot127 commented on librz/util/float/float_internal.c:
```diff
@@ -0,0 +1,440 @@
+
+/**
+ * Shift right, but keeps LSB true if hit 1 during shift
+ * \param x RzBitVector, pointer to bv
```
`\param dist` missing.

Rot127 reviewed (commented):
Nice work! Thank you for all the effort!

Heersin commented:
@Rot127 @DMaroo  Thank you for your reviews, they are helpful. Sorry I could not deal with them immediately. I have to work on my course projects these days (DDL at 12.July), and then I could start to solve these issues and continue the work.

Heersin committed:
Add comments and solve some issues

Heersin committed:
Remove unused functions

b1llow commented on librz/util/float/float.c:
```diff
@@ -1,31 +1,92 @@
+ */
+RZ_API RZ_OWN char *rz_float_bv_as_string(RZ_NULLABLE RzFloat *f) {
+	if (!f || !f->s) {
+		return strdup("NULL");
```
Maybe it should be `NULL` pointer instead of the string `"NULL"`

wargio replied:
i agree.

b1llow commented on librz/util/float/float.c:
```diff
@@ -1,31 +1,92 @@
+	}
 
-static void print_f_bv(RzBitVector *bv, const char *mark) {
+	str[total + 2] = '\0';
```
Seems to be out of bounds

Heersin replied:
yep. ready to remove the entire function though

b1llow commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1534 @@
+ * \param sign sign bit of infinity
+ * \return an infinity float
+ */
+RZ_API RZ_OWN RzFloat *rz_float_new_inf(RzFloatFormat format, bool sign) {
```
I think there can be some special value as static constant.

Such as : +-0, +-inf, nan

wargio commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1534 @@
+	ut32 exp_len = rz_float_get_format_info(f->r, RZ_FLOAT_INFO_EXP_LEN);
+	ut32 total = rz_float_get_format_info(f->r, RZ_FLOAT_INFO_TOTAL_LEN);
+
+	char *str = (char *)malloc(total + 2);
```
+3, otherwise is OOB

wargio commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1534 @@
+ * \return a human-readable string of float.
+ * exponent part and mantissa part would be spilit
+ */
+RZ_API RZ_OWN char *rz_float_as_string(RZ_NONNULL RzFloat *f) {
```
i would call this more `rz_float_as_bit_string`

Heersin replied:
this one would return a string likes `+0111000 | 000000000001111111111...`. 

I prefer to rename `rz_float_bv_as_string` to `rz_float_as_bit_string`, because it returns a pure bit string without extra symbols (`00111000000000000001111111111...`)

wargio replied:
ok, for me.

XVilka replied:
Ok for me as well.

Heersin committed:
Fix bug in float_new and add detect_spec test

Heersin committed:
Fix too big constant

Heersin committed:
Use updated bitvector API

Heersin committed:
Add remainder framework

Heersin committed:
Add remainder implementation according to MPFR

Remain a bug before rounding

Bug in testcase2

Pass testcase 2

Fix the sign in remainder

Add remainder and mod, split from the big rem

Clean code

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1797 @@
+#include "float_internal.c"
+
+/**
+ * \breif return the bitvector string of a float
```
`\brief`

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1797 @@
+ * \param f float
+ * \return a human-readable string of float.
+ * exponent part and mantissa part would be spilit
+ */
```
I recommend adding a few examples on how exponent and mantissa part would look like

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1797 @@
+#define PROC_SPECIAL_FLOAT_END }
+
+/**
+ * Get const attributes from float
```
Probably `\brief`

XVilka commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1797 @@
+}
+
+/**
+ * Init the bv inside float
```
`bitvector`

XVilka reviewed (changes requested):
Looks ready to go, IMHO. We can improve the implementation during adding the RzIL support for Float.

wargio commented:
LGTM too, awesome work

Heersin committed:
Fix typo and other issues

Heersin committed:
Add general arithmetic operation for float and make some API internal

XVilka reviewed (approved):
@Heersin is it still WIP? Looks ready to me.

wargio commented:
```
Mismatched function annotation for rz_float_as_bit_string at <rizin/librz/include/rz_util/rz_float.h:118:21> : RZ_NONNULL (was RZ_NULLABLE at <rizin/librz/util/float/float.c:26:21>
Mismatched function annotation for rz_float_as_hex_string at <rizin/librz/include/rz_util/rz_float.h:119:21> : RZ_NONNULL (was RZ_NULLABLE at <rizin/librz/util/float/float.c:39:21>
Mismatched function annotation for rz_float_abs at <rizin/librz/include/rz_util/rz_float.h:114:24> : RZ_API (was RZ_OWN at <rizin/librz/util/float/float.c:1755:24>
Mismatched function annotation for rz_float_abs at <rizin/librz/include/rz_util/rz_float.h:114:24> : RZ_OWN (was None at <rizin/librz/util/float/float.c:1755:24>
```

Heersin committed:
Fix mismatched fn signature

Heersin commented:
@XVilka I will make it ready after fix the mismatch.
@wargio Thank you wargio :+1:

wargio commented on librz/include/rz_util/rz_float.h:
```diff
@@ -0,0 +1,125 @@
+RZ_API RZ_OWN RzFloat *rz_float_abs(RZ_NONNULL RzFloat *f);
+RZ_API RZ_OWN RzFloat *rz_float_new_from_hex_as_f64(ut64 hex_value);
+RZ_API RZ_OWN RzFloat *rz_float_new_from_hex_as_f32(ut32 hex_value);
+RZ_API RZ_OWN char *rz_float_as_string(RZ_NONNULL RzFloat *f);
+RZ_API RZ_OWN char *rz_float_as_bit_string(RZ_NULLABLE RzFloat *f);
+RZ_API RZ_OWN char *rz_float_as_hex_string(RZ_NULLABLE RzFloat *f, bool use_pad);
```
i just wonder why rz_float_as_string is not nullable while the others are

XVilka replied:
Good catch, yes, it's better to make them consistent.

Heersin replied:
Good catch

wargio commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1884 @@
+	ut32 pos = rz_bv_len(f->s) - 1;
+	ut32 i;
+
+	str[0] = rz_float_get_sign(f) ? '-' : '+';
```
shall we always put + on positive numbers? (do not change this for now)

XVilka replied:
Not sure about this.

Heersin replied:
not necessary for positive numbers.

wargio commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1884 @@
+ * \param hex_value 64-bit hex_value
+ * \return RzFloat-binary64
+ */
+RZ_API RZ_OWN RzFloat *rz_float_new_from_hex_as_f64(ut64 hex_value) {
```
i do not like this name, also what is the endianness?

Heersin replied:
I think it's big-endian. I use the bv_from_ut64 inside it, so an hex value (`0x3dfff7bf`) would be converted to bitvector (`00111101111111111111011110111111`). and then be represented as float, with MSB of bitvector became the sign bit.

wargio commented on librz/util/float/float.c:
```diff
@@ -0,0 +1,1884 @@
+ * \param f float
+ * \return true if it's an infinity, else false
+ */
+RZ_API bool rz_float_is_inf(RZ_NONNULL RzFloat *f) {
```
these RZ_NONNULL are actually not checked.

b1llow reviewed (approved):
awesome works

Heersin committed:
Make as_string functions consistent
